mvpoly/tests/prime.rs

```diff
@@ -453,3 +457,166 @@ fn test_eval_pbt_neg() {
+    }
+
+    impl Interpreter for Constraint {
+        type Position = Column;
+
+        type Variable = Expr<ConstantExpr<Fp>, Column>;
+
+        fn allocate(&mut self) -> Self::Position {
+            let col = Column::X(self.idx);
+            self.idx += 1;
+            col
+        }
+
+        fn fetch(&self, col: Self::Position) -> Self::Variable {
+            Expr::Atom(ExprInner::Cell(Variable {
+                col,
+                row: CurrOrNext::Curr,
+            }))
+        }
+    }
+
+    impl Constraint {
+        fn new() -> Self {
+            Self { idx: 0 }
+        }
+    }
+
+    let mut interpreter = Constraint::new();
+    // Constraints for elliptic curve addition, without handling the case of the
+    // point at infinity or double
+    let lambda = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+    let x1 = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+    let x2 = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+
+    let y1 = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+    let y2 = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+
+    let x3 = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+    let y3 = {
+        let pos = interpreter.allocate();
+        interpreter.fetch(pos)
+    };
+
+    // Check we can convert into a Dense polynomial using prime representation
+    // We have 7 variables, maximum degree 2
+    // We test by evaluating at a random point.
+    let mut rng = o1_utils::tests::make_test_rng(None);
+    {
+        // - Constraint 1: λ (X1 - X2) - Y1 + Y2 = 0
+        let expression = lambda.clone() * (x1.clone() - x2.clone()) - (y1.clone() - y2.clone());
+
+        let p = Dense::<Fp, 7, 2>::from(expression);
+        let random_evaluation: [Fp; 7] = std::array::from_fn(|_| Fp::rand(&mut rng));
+        let eval = p.eval(&random_evaluation);
+        let exp_eval = {
+            random_evaluation[0] * (random_evaluation[1] - random_evaluation[2])
+                - (random_evaluation[3] - random_evaluation[4])
+        };
+        assert_eq!(eval, exp_eval);
+    }
+
+    {
+        // - Constraint 2: X3 + X1 + X2 - λ^2 = 0
+        let expr = x3.clone() + x1.clone() + x2.clone() - lambda.clone() * lambda.clone();
+        let p = Dense::<Fp, 7, 2>::from(expr);
+        let random_evaluation: [Fp; 7] = std::array::from_fn(|_| Fp::rand(&mut rng));
+        let eval = p.eval(&random_evaluation);
+        let exp_eval = {
+            random_evaluation[5] + random_evaluation[1] + random_evaluation[2]
+                - random_evaluation[0] * random_evaluation[0]
+        };
+        assert_eq!(eval, exp_eval);
+    }
+    {
+        // - Constraint 3: Y3 - λ (X1 - X3) - Y1 = 0
+        let expr = y3.clone() - lambda.clone() * (x1.clone() - x3.clone()) + y1.clone();
```

Comment:
equation does not match the expression.

Reply:
Addressed in https://github.com/o1-labs/proof-systems/pull/2505